### PR TITLE
instance_completion_substask

### PR DESCRIPTION
**Pull request by aothms**
Note:

- @rw-bsi please suggest how to tie this into the celery workflow, I didn't really understand chord() this is probably not how we want it
- couldn't really test because the checked in report view doesn't show instances yet

### CODE REVIEW

File: backend/apps/ifc_validation/tasks.py [hunk 1]
```diff
@@ -178,10 +178,25 @@ def ifc_file_validation_task(self, id, file_name, *args, **kwargs):
         chord(
             parallel_tasks,
             workflow_completed
-        ).on_error(chord_error_task))
+        ).on_error(chord_error_task) |
```
Comment (rw-bsi) [inline backend/apps/ifc_validation/tasks.py:181]:
```
    final_tasks = chain(
        instance_completion_subtask.s(id, file_name)
    )

    workflow = (
        workflow_started |
        serial_tasks |
        chord(
            chord(parallel_tasks, final_tasks).on_error(chord_error_task),
            workflow_completed
        ).on_error(chord_error_task))
```

File: backend/apps/ifc_validation/tasks.py [hunk 2]
```diff
@@ -178,10 +178,25 @@ def ifc_file_validation_task(self, id, file_name, *args, **kwargs):
...
+        instance_completion_substask.s(id, file_name))
     workflow.set(link_error=[error_task])
     workflow.apply_async()
 
+@shared_task(bind=True)
```
Comment (rw-bsi) [inline backend/apps/ifc_validation/tasks.py:186]:
```
@shared_task(bind=True)
@log_execution
@requires_django_user_context
def instance_completion_subtask(self, prev_result, id, file_name, *args, **kwargs):

    # fetch request info
    request = ValidationRequest.objects.get(pk=id)
    file_path = get_absolute_file_path(request.file.name)

    # set overall progress
    PROGRESS_INCREMENT = 10
    request.progress = PROGRESS_INCREMENT
    request.save()
    
    # add task
    task = ValidationTask.objects.create(request=request, type=ValidationTask.Type.INSTANCE_COMPLETION)
    task.mark_as_initiated()

    # update Model Instance records
    try:
        
        task.set_process_details(None, f"(module) ifcopenshell.open() for file '{file_path}')")

        ifc_file = ifcopenshell.open(file_path)

        with transaction.atomic():

            counter = 0
            for inst in request.model.instances.iterator():
                counter = counter + 1
                inst.ifc_type = ifc_file[inst.stepfile_id].is_a()
                inst.save()
            
            reason = f'Instance Completion updated {counter:,} Model Instance record(s).'
            task.mark_as_completed(reason)
            return {'is_valid': True, 'reason': reason}
        
    except ifcopenshell.Error as err:
        reason = str(err)
        task.mark_as_completed(reason)
        return {'is_valid': False, 'reason': reason}

    except Exception as err:
            task.mark_as_failed(err)
            raise

```

**Review by rw-bsi — COMMENTED** [review]
See comments in file - that's how I would do it

**Comment by rw-bsi** [comment]
Example result:

image
image

Before:
image

After:
image

Looks good to me then @aothms

**Comment by aothms** [comment]
committed as suggested @rw-bsi, thanks.

**Pull request merged**